web/source/osk/oskManager.ts

```diff
@@ -877,6 +956,36 @@ namespace com.keyman.osk {
       }
     }
 
+    public refreshLayout(): void {
+      // Step 1:  have the necessary conditions been met?
+      const fixedSize = this.width && this.height && this.width.absolute && this.height.absolute;
+      const computedStyle = getComputedStyle(this._Box);
+      const isInDOM = computedStyle.height != '' && computedStyle.height != 'auto';
+
+      // Step 2:  determine basic layout geometry
+      if(fixedSize) {
+        this._computedWidth  = this.width.val;
+        this._computedHeight = this.height.val;
+      } else if(isInDOM) {
+        this._computedWidth   = parseInt(computedStyle.width, 10);
+        this._computedHeight  = parseInt(computedStyle.height, 10);
+      } else {
+        // Cannot perform layout operations!
+        return;
+      }
+
+      // Step 3:  perform layout operations.
+      this.needsLayout = false;
```

Comment:
Once we know that we can properly perform layout calculations, the first thing we do is determine the internal layout so that everything inside the OSK can have its proper dimensions for use within their own `layoutRefresh` operations.

// TODO:  don't forget to do a `vkbd.setSize(this.computedWidth, this.computedHeight)` here!  (_After_ line 978.)